### PR TITLE
KAFKA-5361: Add more integration tests for Streams EOS

### PR DESCRIPTION
**Pull request by mjsax**
- multi-subtopology tests
 - fencing test
 - producer fenced bug fix: Streams did not recover correctly from ProducerFencedException

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @enothereska @dguy @bbejeck @guozhangwang

**Comment by mjsax** [comment]
Call for review @enothereska @dguy @bbejeck @guozhangwang

**Comment by mjsax** [comment]
Test failure unrelated: `kafka.security.auth.SimpleAclAuthorizerTest.testHighConcurrencyModificationOfResourceAcls`

**Comment by mjsax** [comment]
Retest this please

**Review by bbejeck — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java [hunk 1]
```diff
@@ -57,7 +56,7 @@ public static void registerAppInfo(String prefix, String id) {
             AppInfo mBean = new AppInfo();
             ManagementFactory.getPlatformMBeanServer().registerMBean(mBean, name);
         } catch (JMException e) {
-            log.warn("Error registering AppInfo mbean", e);
+            log.warn("Error registering AppInfo mbean: {}", e.getMessage());
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java]: Why this change?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -259,7 +260,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated, f
                     } catch (final LockException e) {
                         // ignore and retry
                         if (!retryingTasks.contains(taskId)) {
-                            log.warn("{} Could not create task {}. Will retry. {}", logPrefix, taskId, e);
+                            log.warn("{} Could not create task {}. Will retry. {}", logPrefix, taskId, e.getMessage());
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: `could not create task {} due to {}. Will retry.`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -272,6 +273,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated, f
                 try {
                     Thread.sleep(backoffTimeMs);
                     backoffTimeMs <<= 1;
+                    backoffTimeMs = Math.min(backoffTimeMs, maxBackoffTimeMs);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Why this change again? Or could you add this to the description of the PR (it seems not included)?

**Review by guozhangwang — COMMENTED** [review]
Left some minor comments.

Could you trigger say 5 consecutive builds on Jenkins and check their results?

File: clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java (hunk 1 above)
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java]: I though we don't want to print stack traces for WARN level logs?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Oh. My bad. That is unrelated and I included it to not forget about it -- currently, with exponential increase of waiting time users did report unsatisfying behavior for this. I think we need to fix it... This should also be back ported to `0.10.2`... But maybe it should not be part of this PR. But not sure how else we can get it into `0.11` ;)

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Opened #3312 for `0.10.2`

**Comment by mjsax** [comment]
Updated.

File: clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java]: There is no such policy at the moment. Was there a specific issue with this line?

Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java]: The issue we observed in Streams is, that people see a stack trace and get worried -- they don't pay attention to the log level that is `WARN` -- thus it seems better to never include stack traces on `WARN` to not confuse users. I can revert, too.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Hmm.. I was not aware of this, and could you point me to the un-statisfying behavior that user complain about? Would like to make sure if this is really an issue to fix before merging.

For this PR, I'd suggest we do not include it so that it can be merged.

File: clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java]: `WARN` is something to be worried about, right? If it's just informational, it should be `INFO`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Reverted this and created https://github.com/apache/kafka/pull/3327

**Comment by guozhangwang** [comment]
LGTM and merged to trunk / 0.11.0.

File: clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java (hunk 1 above)
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java]: I agree. But `WARN` should still not contain a stack trace -- that's for `ERROR` IMHO.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/utils/AppInfoParser.java]: That seems pretty arbitrary though since you just agreed that users should be worried either way. The problem may be that people are using warn when they should use info.
